spec.html

```diff
@@ -36005,27 +36009,43 @@ <h1>
             1. Assert: The mathematical value of _A_'s *"length"* property is _n_ + 1.
             1. Perform ! CreateDataPropertyOrThrow(_A_, *"index"*, 𝔽(_lastIndex_)).
             1. Perform ! CreateDataPropertyOrThrow(_A_, *"input"*, _S_).
-            1. Let _matchedSubstr_ be the substring of _S_ from _lastIndex_ to _e_.
+            1. Let _match_ be the Match Record { [[StartIndex]]: _lastIndex_, [[EndIndex]]: _e_ }.
+            1. Let _indices_ be a new empty List.
+            1. Let _groupNames_ be a new empty List.
+            1. Append _match_ to _indices_.
+            1. Let _matchedSubstr_ be ! GetMatchString(_S_, _match_).
             1. Perform ! CreateDataPropertyOrThrow(_A_, *"0"*, _matchedSubstr_).
             1. If _R_ contains any |GroupName|, then
               1. Let _groups_ be ! OrdinaryObjectCreate(*null*).
+              1. Let _hasGroups_ be *true*.
             1. Else,
               1. Let _groups_ be *undefined*.
+              1. Let _hasGroups_ be *false*.
             1. Perform ! CreateDataPropertyOrThrow(_A_, *"groups"*, _groups_).
             1. For each integer _i_ such that _i_ &ge; 1 and _i_ &le; _n_, in ascending order, do
               1. Let _captureI_ be _i_<sup>th</sup> element of _r_'s _captures_ List.
-              1. If _captureI_ is *undefined*, let _capturedValue_ be *undefined*.
-              1. Else if _fullUnicode_ is *true*, then
-                1. Assert: _captureI_ is a List of code points.
-                1. Let _capturedValue_ be ! CodePointsToString(_captureI_).
+              1. If _captureI_ is *undefined*, then
+                1. Let _capturedValue_ be *undefined*.
+                1. Append *undefined* to _indices_.
               1. Else,
-                1. Assert: _fullUnicode_ is *false*.
-                1. Assert: _captureI_ is a List of code units.
-                1. Let _capturedValue_ be the String value consisting of the code units of _captureI_.
+                1. Let _captureStart_ be _captureI_'s _startIndex_.
+                1. Let _captureEnd_ be _captureI_'s _endIndex_.
+                1. If _fullUnicode_ is *true*, then
+                  1. Set _captureStart_ to ! GetStringIndex(_S_, _Input_, _captureStart_).
+                  1. Set _captureEnd_ to ! GetStringIndex(_S_, _Input_, _captureEnd_).
```

Comment:
In reference to possibly dropping GetStringIndex, we could just do this instead:

```suggestion
                  1. _captureStart_ is an index into the _Input_ character list, derived from _S_, matched by _matcher_. Let _captureStartUTF_ be the smallest index into _S_ that corresponds to the character at element _captureStart_ of _Input_. If _captureStart_ is greater than or equal to the number of elements in _Input_, then _captureStartUTF_ is the number of code units in _S_.
                  1. Set _captureStart_ to _captureStartUTF_.
                  1. _captureEnd_ is an index into the _Input_ character list, derived from _S_, matched by _matcher_. Let _captureStartUTF_ be the smallest index into _S_ that corresponds to the character at element _captureEnd_ of _Input_. If _captureEnd_ is greater than or equal to the number of elements in _Input_, then _captureStartUTF_ is the number of code units in _S_.
                  1. Set _captureEnd_ to _captureEndUTF_.
```